internal/handlers/pg/msg_create.go

```diff
@@ -69,21 +71,26 @@ func (h *Handler) MsgCreate(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg,
 		return nil, err
 	}
 
-	if err := h.pgPool.CreateDatabase(ctx, db); err != nil && !errors.Is(err, pgdb.ErrAlreadyExist) {
-		return nil, lazyerrors.Error(err)
-	}
-
-	// TODO use a transaction instead of pgPool: https://github.com/FerretDB/FerretDB/issues/866
-	if err = h.pgPool.CreateCollection(ctx, h.pgPool, db, collection); err != nil {
-		if errors.Is(err, pgdb.ErrAlreadyExist) {
-			msg := fmt.Sprintf("Collection already exists. NS: %s.%s", db, collection)
-			return nil, common.NewErrorMsg(common.ErrNamespaceExists, msg)
+	err = h.pgPool.InTransaction(ctx, func(tx pgx.Tx) error {
+		if err := pgdb.CreateDatabaseIfNotExists(ctx, tx, db); err != nil {
+			return lazyerrors.Error(err)
 		}
-		if errors.Is(err, pgdb.ErrInvalidTableName) {
-			msg := fmt.Sprintf("Invalid collection name: '%s.%s'", db, collection)
-			return nil, common.NewErrorMsg(common.ErrInvalidNamespace, msg)
+
+		if err = pgdb.CreateCollection(ctx, tx, db, collection); err != nil {
```

Comment:
```suggestion
		if err := pgdb.CreateCollection(ctx, tx, db, collection); err != nil {
```

Otherwise, we are assigning to the outer `err`.

Maybe there is a more general solution? For example, start each closure with `var err error`?

Reply:
No comment about more general solution?